command/agent/template/template_test.go

```diff
@@ -0,0 +1,156 @@
+	ts := httptest.NewServer(http.HandlerFunc(handleRequest))
+	defer ts.Close()
+	tmpDir, err := ioutil.TempDir("", "agent-tests")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testCases := map[string]struct {
+		templates []*ctconfig.TemplateConfig
+	}{
+		"basic": {
+			templates: []*ctconfig.TemplateConfig{
+				&ctconfig.TemplateConfig{
+					Contents: testhelpers.StrPtr(templateContents),
+				},
+			},
+		},
+	}
+
+	templateTokenCh := make(chan string, 1)
+	var unblockCh <-chan struct{}
+
+	// secretRender is a simple struct that represents the secret we render to
+	// disk. It's used to unmarshal the file contents and test against
+	type secretRender struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+		Version  string `json:"version"`
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			for i, template := range tc.templates {
+				// template.Destination = testhelpers.StrPtr(fmt.Sprintf("%s/render_%d.txt", tmpDir, i))
+				dstFile := fmt.Sprintf("%s/render_%d.txt", tmpDir, i)
+				template.Destination = testhelpers.StrPtr(dstFile)
+
+				ctx, cancelFunc := context.WithCancel(context.Background())
+				sc := ServerConfig{
+					Logger: logging.NewVaultLogger(hclog.Trace),
+					VaultConf: &config.Vault{
+						Address: ts.URL,
+					},
+				}
+
+				var server *Server
+				server, unblockCh = NewServer(&sc)
+				if ts == nil {
+					t.Fatal("nil server returned")
+				}
+				if unblockCh == nil {
+					t.Fatal("nil blocking channel returned")
+				}
+
+				go server.Run(ctx, templateTokenCh, tc.templates)
+
+				// send a dummy value to trigger the internal Runner to query for secret
+				// info
+				templateTokenCh <- "test"
+
+				// Unblock should close immediately b/c there are no templates to render
+				select {
+				case <-ctx.Done():
+				case <-unblockCh:
+				}
+
+				// cancel to clean things up
+				cancelFunc()
+
+				// verify test file exists and has the content we're looking for
+				content, err := ioutil.ReadFile(dstFile)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				secret := secretRender{}
+				if err := json.Unmarshal(content, &secret); err != nil {
+					panic(err)
+				}
+				if secret.Username != "appuser" || secret.Password != "password" || secret.Version != "3" {
+					t.Fatalf("secret didn't match: %#v", secret)
+				}
+			}
+		})
+	}
+}
+
+func handleRequest(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintln(w, jsonResponse)
+}
+
+var jsonResponse = `
+{
+  "request_id": "8af096e9-518c-7351-eff5-5ba20554b21f",
+  "lease_id": "",
+  "renewable": false,
+  "lease_duration": 0,
+  "data": {
+    "data": {
+      "password": "password",
+      "username": "appuser"
+    },
+    "metadata": {
+      "created_time": "2019-10-07T22:18:44.233247Z",
+      "deletion_time": "",
+      "destroyed": false,
+      "version": 3
+    }
+  },
+  "wrap_info": null,
+  "warnings": null,
+  "auth": null
+}
+`
+
+var templateContents = `
```

Comment:
I anticipate this simple template being reused across other tests, so I'd like to leave it here for now. Are there significant drawbacks to sharing inputs across tests that I should be aware of, in the context of rendering this simple template?

Reply:
It's definitely an opinion I have which is not universal across the team.  But I've seen plenty of test suites in my life that followed that pattern, which over the years became really hard to follow as a result.  The shared input tends to be overcomplicated in order to cater to each test's needs, and the problem only gets worse with time.   You can't really tell without great effort what each test depends on in their shared input.   Because the shared input is liable to contain things not needed by a given test, it reduces the clarity of what the test is actually doing.  And because it's unclear just what's needed for a minimal test, developers writing new tests are incentivized to double down and further exacerbate the problem.